### PR TITLE
[SPARK-29497][CONNECT] Throw error when UDF is not deserializable.

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds a better error message when a JVM UDF cannot be deserialized.

### Why are the changes needed?
In some cases a UDF cannot be deserialized. The happens when a lambda references itself (typically through the capturing class). Java cannot deserialize such an object graph because SerializedLambda's are serialization proxies which need the full graph to be deserialized before they can be transformed into the actual lambda. This is not possible if there is such a cycle. This PR adds a more readable and understandable error when this happens, the original java one is a `ClassCastException` which does not explain or give any hints of what is actually going on.

### Does this PR introduce _any_ user-facing change?
Yes. It will throw an error on the client when a UDF is not deserializable. The error is better and more actionable then what we got before.

### How was this patch tested?
Added tests.

### CODE REVIEW

**Review by zhenlineo — APPROVED** [review]
LGTM

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala [hunk 1]
```diff
@@ -144,6 +146,25 @@ case class ScalarUserDefinedFunction private[sql] (
...
+
+  private def checkDeserializable(bytes: Array[Byte]): Unit = {
+    try {
+      SparkSerDeUtils.deserialize(bytes, SparkClassUtils.getContextOrSparkClassLoader)
+    } catch {
+      case e: ClassCastException if e.getMessage.contains(LAMBDA_DESERIALIZATION_ERR_MSG) =>
+        throw new SparkException(
+          "UDF cannot be executed on a Spark cluster, it cannot be deserialized." +
+            "This is very likely to be caused by the UDF having a self-reference, " +
+            "this is not supported by java serialization.")
```
Comment (amaliujia) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala:160]:
```suggestion
        throw new SparkException(
          "UDF cannot be executed on a Spark cluster: it cannot be deserialized." +
            "This is very likely to be caused by the UDF having a self-reference. " +
            "This is not supported by java serialization.")
```

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala [hunk 2]
```diff
@@ -144,6 +146,25 @@ case class ScalarUserDefinedFunction private[sql] (
...
+      case NonFatal(e) =>
+        throw new SparkException(
+          "UDF cannot be executed on a Spark cluster, it cannot be deserialized.",
+          e)
```
Comment (amaliujia) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala:164]:
```suggestion
        throw new SparkException(
          "UDF cannot be executed on a Spark cluster: it cannot be deserialized.",
          e)
```

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala [hunk 3]
```diff
@@ -144,6 +146,25 @@ case class ScalarUserDefinedFunction private[sql] (
...
+    "cannot assign instance of java.lang.invoke.SerializedLambda to field"
+
+  private def checkDeserializable(bytes: Array[Byte]): Unit = {
+    try {
+      SparkSerDeUtils.deserialize(bytes, SparkClassUtils.getContextOrSparkClassLoader)
+    } catch {
+      case e: ClassCastException if e.getMessage.contains(LAMBDA_DESERIALIZATION_ERR_MSG) =>
+        throw new SparkException(
+          "UDF cannot be executed on a Spark cluster: it cannot be deserialized. " +
+            "This is very likely to be caused by the lambda function (the UDF) having a "
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala:159]:
```suggestion
            "This is very likely to be caused by the lambda function (the UDF) having a " +
```

Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala:159]: It seems better to use `"""..."""`.

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala:159]: @beliefer that would create a multi-line string right?

**Review by LuciferYang — APPROVED** [review]

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala (hunk 3 above)
Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala]: We can stripe it if we do not need multi-line string. Just for the code style.

**Review by amaliujia — APPROVED** [review]

**Comment by hvanhovell** [comment]
Merging. Error is unrelated.

**Pull request closed without merging**